rules/manager.go

```diff
@@ -969,6 +975,22 @@ func (m *Manager) Update(interval time.Duration, files []string, externalLabels
 	return nil
 }
 
+// GroupLoader is responsible for loading rule groups from arbitrary sources and parsing them.
+type GroupLoader interface {
+	Load(identifier string) (*rulefmt.RuleGroups, []error)
+	Parse(query string) (fmt.Stringer, error)
+}
+
+// FileLoader is the default GroupLoader implementation. It defers to rulefmt.ParseFile
+// and parser.ParseExpr
+type FileLoader struct{}
+
+func (FileLoader) Load(identifier string) (*rulefmt.RuleGroups, []error) {
+	return rulefmt.ParseFile(identifier)
+}
+
+func (FileLoader) Parse(query string) (fmt.Stringer, error) { return parser.ParseExpr(query) }
```

Comment:
I would expect a function called "Parse" to parse - not to return a string.

Reply:
the Parse function went from returning a `parser.Expr` to a `fmt.Stringer`. It still parses the query -- I've just narrowed the interface. I did this for two reasons:
1) Because the parsed expression is only used for it's `String()` method. Narrowing this interface is easy and better reflects how the return value is used.
2) It allows consumers of this interface to substitute other underlying types asides `parser.Expr` that implement `fmt.Stringer`.

Reply:
Why are we converting a string to a string?

Reply:
We're not; rather we're converting a string to a type which implements `fmt.Stringer`, which is still a valuable parsing step to ensure the query is valid. Previously this converted a `string` to a `parser.Expr`, but _only the `String` method was used_. This is similar to reducing `io.ReadWriter` to `io.Reader` when only the `Read` method is used. Does this make sense? We're still parsing the string, but now the parsing step is substitutable and doesn't bind the output type to `parser.Expr`.

Reply:
For some additional background (pulled from the initial PR description),
> This change also enables other non-default use cases in downstream projects such as parsing LogQL queries in Loki.

Reply:
Poking at the code, that we only use String is a long standing wart that I thought we had already fixed by now. This type should stay as-is, and we should not be doing an unnecessary round-trip through the parser here.

Reply:
Can you elaborate on 
> we should not be doing an unnecessary round-trip through the parser here.


As far as I can tell, this isn't an unnecessary round trip as there is no other validation of the query. That being said, I'll set it back to `parser.Expr`.

Reply:
We parse the query and then render it again, and then parse the rendered value when we run the rule. If we want to do just validation, it shouldn't involve a return value or that rendering step.